### PR TITLE
Added Attack classes

### PR DESCRIPTION
**Pull request by abheesht17**
I have added the attack classes.
Please see if they are okay.

I have not added integration tests. Can someone help me with those?

### CODE REVIEW

**Comment by rajaswa** [comment]
This might take some time to get reviewed. Adding @Sharad24  , @someshsingh22  and @shashank-m  as reviewers too

File: decepticonlp/attacks/attack.py [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+loss_type_list = [
+    torch.nn.modules.loss.BCELoss,
+    torch.nn.modules.loss.BCEWithLogitsLoss,
+    torch.nn.modules.loss.CrossEntropyLoss,
+    torch.nn.modules.loss.NLLLoss,
+]
+
+
+class CharAttackerWithHuggingFace:
+    """
```
Comment (Sharad24) [inline decepticonlp/attacks/attack.py]: Specify here what kind of model it can work with? As in, something like NMT model which takes as input ... and outputs ...

File: decepticonlp/attacks/attack.py [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+        input_format,
+        accuracy=True,
+        logs_after_every=50,
+        device=torch.device("cpu"),
+    ):
+
+        # initialise parameters
+        self.model = model.to(device)
+        self.data_loader = data_loader
+        self.adversarial_data_loader = adversarial_data_loader
```
Comment (Sharad24) [inline decepticonlp/attacks/attack.py]: Why do we need to have 2 data loaders? 

cc: @rajaswa

File: decepticonlp/attacks/attack.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+                    type(self.criterion["loss"])
+                    == torch.nn.modules.loss.BCEWithLogitsLoss
+                ):
+                    computed_labels = (outputs > self.threshold).float()
+
+                elif (
+                    type(self.criterion["loss"]) == torch.nn.modules.loss.NLLLoss
+                ) or (
+                    type(self.criterion["loss"])
+                    == torch.nn.modules.loss.CrossEntropyLoss
```
Comment (Sharad24) [inline decepticonlp/attacks/attack.py]: The correct way to check these things for an if statement is to use `isinstance(self.criterion["loss"], torch.nn.modules.loss.CrossEntropyLoss)`

File: decepticonlp/attacks/attack.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+		adversarial_data_loader: torch.utils.data.Dataloader
+			-Dataloader object of the adversarial test set
+		
+		input_format: list (of strings) (default: ['input_ids','labels'])
+			-The input format as you would feed to your model.
+			Eg:
+				For BERTForSequenceClassification, the input_format will be ['input_ids','attention_masks','labels'] (if you are feeding only these three to the model).
+		
+		accuracy: boolean (default: True)
+			-If True, logs and prints the accuracy
```
Comment (Sharad24) [inline decepticonlp/attacks/attack.py]: Also what docstring format is this?

File: decepticonlp/attacks/attack.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+        return loss_logs, accuracy_logs
+
+    # Getter functions
+
+    def get_criterion_logs(self):
+        return self.loss_logs, self.accuracy_logs
+
+
+class CharAttacker:
+    """
```
Comment (Sharad24) [inline decepticonlp/attacks/attack.py]: Need to add a description of what the class does

**Review by Sharad24 — COMMENTED** [review]
Looks good

File: decepticonlp/attacks/attack.py (hunk 1 above)
Comment (abheesht17) [reply decepticonlp/attacks/attack.py]: Will do

File: decepticonlp/attacks/attack.py (hunk 2 above)
Comment (abheesht17) [reply decepticonlp/attacks/attack.py]: One for the original dataset, and the other for the adversarial dataset (the user will apply transforms). I discussed this with Rajaswa and we thought that instead of passing dataloaders, we could pass the Dataset objects. 

However, during implementation, I came across multiple examples online where users have used torchtext iterators, and passing them through these classes would also work, as they are handled the exact same way as dataloaders. So, it comes down to flexibility, which is why I chose passing dataloaders over passing Dataset objects.

File: decepticonlp/attacks/attack.py (hunk 3 above)
Comment (abheesht17) [reply decepticonlp/attacks/attack.py]: Will edit.

File: decepticonlp/attacks/attack.py (hunk 4 above)
Comment (abheesht17) [reply decepticonlp/attacks/attack.py]: https://github.com/pytorch/vision/blob/master/torchvision/transforms/transforms.py
This is what I followed

File: decepticonlp/attacks/attack.py (hunk 5 above)
Comment (abheesht17) [reply decepticonlp/attacks/attack.py]: On it

**Comment by abheesht17** [comment]
@rajaswa Have a look. Instead of two classes, I've defined two functions inside the same class. Also attached example scripts.

**Comment by rajaswa** [comment]
Why do we have a specific charattack class?
And why the special attention to huggingface?

**Comment by rajaswa** [comment]
What we need is one single attacker class.

**Comment by abheesht17** [comment]
@rajaswa  Required changes made.
